packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol

```diff
@@ -0,0 +1,1167 @@
+                abi.encode(_args.proxyAdmin)
+            );
+
+            // ChugSplashProxy requires setting the proxy type on the ProxyAdmin.
+            _args.proxyAdmin.setProxyType(ret, IProxyAdmin.ProxyType.CHUGSPLASH);
+        } else if (LibString.eq(_contractName, "L1CrossDomainMessenger")) {
+            // L1CrossDomainMessenger is a special case ResolvedDelegateProxy (legacy).
+            string memory l1XdmName = "OVM_L1CrossDomainMessenger";
+            ret = Blueprint.deployFrom(
+                blueprints().resolvedDelegateProxy,
+                _computeSalt(_args.l2ChainId, _args.saltMixer, "L1CrossDomainMessenger"),
+                abi.encode(_args.addressManager, l1XdmName)
+            );
+
+            // ResolvedDelegateProxy requires setting the proxy type on the ProxyAdmin.
+            _args.proxyAdmin.setProxyType(ret, IProxyAdmin.ProxyType.RESOLVED);
+            _args.proxyAdmin.setImplementationName(ret, l1XdmName);
+        } else {
+            // Otherwise this is a normal proxy.
+            ret = Blueprint.deployFrom(
+                blueprints().proxy,
+                _computeSalt(_args.l2ChainId, _args.saltMixer, _contractName),
+                abi.encode(_args.proxyAdmin)
+            );
+        }
+
+        // Emit the proxy creation event.
+        emit ProxyCreation(_contractName, ret);
+
+        // Return the final deployment result.
+        return payable(ret);
+    }
+
+    /// @notice Upgrades a contract by resetting the initialized slot and calling the initializer.
+    /// @param _proxyAdmin The proxy admin of the contract.
+    /// @param _target The target of the contract.
+    /// @param _implementation The implementation of the contract.
+    /// @param _data The data to call the initializer with.
+    function _upgrade(IProxyAdmin _proxyAdmin, address _target, address _implementation, bytes memory _data) internal {
+        _upgrade(_proxyAdmin, _target, _implementation, _data, bytes32(0), 0);
+    }
+
+    /// @notice Upgrades a contract by resetting the initialized slot and calling the initializer.
+    /// @param _proxyAdmin The proxy admin of the contract.
+    /// @param _target The target of the contract.
+    /// @param _implementation The implementation of the contract.
+    /// @param _data The data to call the initializer with.
+    /// @param _slot The slot where the initialized value is located.
+    /// @param _offset The offset of the initializer value in the slot.
+    function _upgrade(
+        IProxyAdmin _proxyAdmin,
+        address _target,
+        address _implementation,
+        bytes memory _data,
+        bytes32 _slot,
+        uint8 _offset
+    )
+        internal
+    {
+        // Check to make sure that we're not downgrading. Downgrades aren't inherently dangerous
+        // but we also don't test for them so we don't really know if a specific downgrade will be
+        // dangerous or not. It's easier to just revert instead.
```

Comment:
Should also make clear that this does allow an "upgrade" to the same version, if we want to keep that.